*** C:/Users/dmark/Downloads/rascsi147/src/raspberrypi/ctapdriver.h	Sat Apr 11 22:53:28 2020
--- C:/Users/dmark/Downloads/rascsi152/src/raspberrypi/netdriver.h	Fri Mar 26 21:18:40 2021
***************
*** 1,31 ****
  //---------------------------------------------------------------------------
  //
  //	SCSI Target Emulator RaSCSI (*^..^*)
  //	for Raspberry Pi
- //
  //	Powered by XM6 TypeG Technology.
- //	Copyright (C) 2016-2020 GIMONS
  //
  //	Imported NetBSD support and some optimisation patch by Rin Okuyama.
  //
! //	[ TAPドライバ ]
  //
  //---------------------------------------------------------------------------
  
! #if !defined(ctapdriver_h)
! #define ctapdriver_h
  
! #ifndef ETH_FRAME_LEN
! #define ETH_FRAME_LEN 1514
! #endif
  
  //===========================================================================
  //
  //	Tapドライバ
  //
  //===========================================================================
  class CTapDriver
  {
  public:
  	// 基本ファンクション
  	CTapDriver();
--- 1,69 ----
  //---------------------------------------------------------------------------
  //
  //	SCSI Target Emulator RaSCSI (*^..^*)
  //	for Raspberry Pi
  //	Powered by XM6 TypeG Technology.
  //
+ //	Copyright (C) 2016-2021 GIMONS(Twitter:@kugimoto0715)
  //	Imported NetBSD support and some optimisation patch by Rin Okuyama.
  //
! //	[ Network driver ]
  //
  //---------------------------------------------------------------------------
  
! #if !defined(cnetdriver_h)
! #define cnetdriver_h
  
! class CTapDriver;
! class NetDriver
! {
! public:
! 	// Base functions
! 	NetDriver();
! 										// Constructor
! 	virtual ~NetDriver();
! 										// Destructor
! 
! 	int FASTCALL IsNetEnable() { return m_bTapEnable; }
! 										// Get network functionality enable
! 	int FASTCALL Process(BOOL read, int func, int phase, int len, BYTE *buf);
! 										// Execute
! 	int FASTCALL GetMacAddr(BYTE *buf);
! 										// Get MAC address
! 	void FASTCALL SetMacAddr(BYTE *buf);
! 										// Set MAC address
! 	void FASTCALL ReceivePacket();
! 										// Receive packet
! 	void FASTCALL GetPacketBuf(BYTE *buf);
! 										// Get packet
! 	void FASTCALL SendPacket(BYTE *buf, int len);
! 										// Send packet
! 
! 	CTapDriver *tap;
! 										// TAP driver
! 	BOOL m_bTapEnable;
! 										// TAP enable flag
! 	BYTE mac_addr[6];
! 										// MAC address
! 	int packet_len;
! 										// Receive packet size
! 	BYTE packet_buf[0x1000];
! 										// Receive packet buffer
! 	BOOL packet_enable;
! 										// Receive packet enable
! };
  
  //===========================================================================
  //
  //	Tapドライバ
  //
  //===========================================================================
+ #ifndef ETH_FRAME_LEN
+ #define ETH_FRAME_LEN 1514
+ #endif
+ 
  class CTapDriver
  {
  public:
  	// 基本ファンクション
  	CTapDriver();
***************
*** 48,53 ****
  										// 送信有効フラグ
  	int m_hTAP;
  										// ディスクプリタ
  };
  
! #endif	// ctapdriver_h
--- 86,91 ----
  										// 送信有効フラグ
  	int m_hTAP;
  										// ディスクプリタ
  };
  
! #endif	// cnetdriver_h
